### PR TITLE
Add Grafana dashboard for visualizing deny connections

### PR DESCRIPTION
**Pull request by heanlan**
1. Add source files of grafana-chord-plugin
  2. Refactor networkpolicy_dashboard by adding support for visualizing deny
  connections
  3. Add corresponding documentation
  
  Signed-off-by: heanlan <hanlan@vmware.com>

### CODE REVIEW

**Comment by heanlan** [comment]
Post the queries used in each panel here to assist review

Chord panel
```sql
SELECT CONCAT(sourcePodNamespace, '/', sourcePodName) as srcPod,
CONCAT(destinationPodNamespace, '/', destinationPodName) as dstPod,
sourceTransportPort as srcPort,
destinationTransportPort as dstPort,
destinationServicePort as dstSvcPort,
destinationServicePortName as dstSvc,
destinationIP as dstIP,
SUM(octetDeltaCount) as bytes,
SUM(reverseOctetDeltaCount) as revBytes,
egressNetworkPolicyName,
egressNetworkPolicyRuleAction,
ingressNetworkPolicyName,
ingressNetworkPolicyRuleAction
from flows_policy_view
WHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND $__timeFilter(flowEndSeconds)
GROUP BY srcPod, dstPod, srcPort, dstPort, dstSvcPort, dstSvc, dstIP, egressNetworkPolicyName, egressNetworkPolicyRuleAction, ingressNetworkPolicyName, ingressNetworkPolicyRuleAction
HAVING bytes > 0
order by srcPod
```
Donut panels
```sql
SELECT SUM(octetDeltaCount), 
CASE WHEN ingressNetworkPolicyNamespace != '' THEN CONCAT(ingressNetworkPolicyNamespace, '/', ingressNetworkPolicyName)
ELSE ingressNetworkPolicyName
END AS np
FROM flows_policy_view
WHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND ingressNetworkPolicyName != ''
AND $__timeFilter(flowEndSeconds)
GROUP BY np
HAVING SUM(octetDeltaCount) != 0

SELECT SUM(octetDeltaCount),
CASE WHEN egressNetworkPolicyNamespace != '' THEN CONCAT(egressNetworkPolicyNamespace, '/', egressNetworkPolicyName)
ELSE egressNetworkPolicyName
END AS np
FROM flows_policy_view
WHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND egressNetworkPolicyName != ''
AND $__timeFilter(flowEndSeconds)
GROUP BY np
HAVING SUM(octetDeltaCount) > 0
```

Line Graphs
```sql
SELECT $__timeInterval(flowEndSeconds) as time,
CASE WHEN sourceTransportPort != 0 THEN CONCAT(sourcePodNamespace, '/', sourcePodName, ':', CAST(sourceTransportPort as VARCHAR))
ELSE CONCAT(sourcePodNamespace, '/', sourcePodName)
END AS src,
CASE WHEN destinationServicePortName != '' AND destinationServicePort != 0 THEN CONCAT(destinationServicePortName, ':', CAST(destinationServicePort as VARCHAR))
WHEN destinationServicePortName != '' AND destinationServicePort == 0 THEN destinationServicePortName
WHEN destinationPodName != '' AND destinationTransportPort != 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName, ':', CAST(destinationTransportPort as VARCHAR))
WHEN destinationPodName != '' AND destinationTransportPort == 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName)
ELSE destinationIP
END AS dst,
CASE WHEN ingressNetworkPolicyNamespace != '' THEN CONCAT(ingressNetworkPolicyNamespace,'/', ingressNetworkPolicyName)
ELSE ingressNetworkPolicyName
END AS np,
CONCAT(src, ' -> ', dst, ' : ', np) as pair,
AVG(throughput)
FROM flows_policy_view
WHERE $__timeFilter(flowEndSeconds)
AND sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND ingressNetworkPolicyRuleAction == 1
AND egressNetworkPolicyRuleAction NOT IN (2, 3)
GROUP BY time, src, dst, np
HAVING AVG(throughput) > 0
ORDER BY time

SELECT $__timeInterval(flowEndSeconds) as time,
CASE WHEN sourceTransportPort != 0 THEN CONCAT(sourcePodNamespace, '/', sourcePodName, ':', CAST(sourceTransportPort as VARCHAR))
ELSE CONCAT(sourcePodNamespace, '/', sourcePodName)
END AS src,
CASE WHEN destinationServicePortName != '' AND destinationServicePort != 0 THEN CONCAT(destinationServicePortName, ':', CAST(destinationServicePort as VARCHAR))
WHEN destinationServicePortName != '' AND destinationServicePort == 0 THEN destinationServicePortName
WHEN destinationPodName != '' AND destinationTransportPort != 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName, ':', CAST(destinationTransportPort as VARCHAR))
WHEN destinationPodName != '' AND destinationTransportPort == 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName)
ELSE destinationIP
END
AS dst,
CASE WHEN egressNetworkPolicyNamespace != '' THEN CONCAT(egressNetworkPolicyNamespace,'/', egressNetworkPolicyName)
ELSE egressNetworkPolicyName
END
AS np,
CONCAT(src, ' -> ', dst, ' : ', np) as pair,
AVG(throughput)
FROM flows_policy_view
WHERE $__timeFilter(flowEndSeconds)
AND sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND egressNetworkPolicyRuleAction == 1
AND ingressNetworkPolicyRuleAction not in (2, 3)
GROUP BY time, src, dst, np
HAVING AVG(throughput) > 0
ORDER BY time

SELECT $__timeInterval(flowEndSeconds) as time,
CASE WHEN sourceTransportPort != 0 THEN CONCAT(sourcePodNamespace, '/', sourcePodName, ':', CAST(sourceTransportPort as VARCHAR))
ELSE CONCAT(sourcePodNamespace, '/', sourcePodName)
END AS src,
CASE WHEN destinationServicePortName != '' AND destinationServicePort != 0 THEN CONCAT(destinationServicePortName, ':', CAST(destinationServicePort as VARCHAR))
WHEN destinationServicePortName != '' AND destinationServicePort == 0 THEN destinationServicePortName
WHEN destinationPodName != '' AND destinationTransportPort != 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName, ':', CAST(destinationTransportPort as VARCHAR))
WHEN destinationPodName != '' AND destinationTransportPort == 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName)
ELSE destinationIP
END
AS dst,
CASE WHEN ingressNetworkPolicyNamespace != '' THEN CONCAT(ingressNetworkPolicyNamespace,'/', ingressNetworkPolicyName)
ELSE ingressNetworkPolicyName
END
AS np,
CONCAT(src, ' -> ', dst, ' : ', np) as pair,
AVG(throughput)
FROM flows_policy_view
WHERE $__timeFilter(flowEndSeconds)
AND sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND ingressNetworkPolicyRuleAction in (2,3)
GROUP BY time, src, dst, np
HAVING AVG(throughput) > 0
ORDER BY time

SELECT $__timeInterval(flowEndSeconds) as time,
CASE WHEN sourceTransportPort != 0 THEN CONCAT(sourcePodNamespace, '/', sourcePodName, ':', CAST(sourceTransportPort as VARCHAR))
ELSE CONCAT(sourcePodNamespace, '/', sourcePodName)
END AS src,
CASE WHEN destinationServicePortName != '' AND destinationServicePort != 0 THEN CONCAT(destinationServicePortName, ':', CAST(destinationServicePort as VARCHAR))
WHEN destinationServicePortName != '' AND destinationServicePort == 0 THEN destinationServicePortName
WHEN destinationPodName != '' AND destinationTransportPort != 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName, ':', CAST(destinationTransportPort as VARCHAR))
WHEN destinationPodName != '' AND destinationTransportPort == 0 THEN CONCAT(destinationPodNamespace,'/', destinationPodName)
ELSE destinationIP
END
AS dst,
CASE WHEN egressNetworkPolicyNamespace != '' THEN CONCAT(egressNetworkPolicyNamespace,'/', egressNetworkPolicyName)
ELSE egressNetworkPolicyName
END
AS np,
CONCAT(src, ' -> ', dst, ' : ', np) as pair,
AVG(throughput)
FROM flows_policy_view
WHERE $__timeFilter(flowEndSeconds)
AND sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')
AND egressNetworkPolicyRuleAction in (2,3)
GROUP BY time, src, dst, np
HAVING AVG(throughput) > 0
ORDER BY time
```

File: docs/network-flow-visibility.md [hunk 1]
```diff
@@ -387,11 +387,36 @@ and throughput is shown in the line graphs.
...
+time range. We use green color ![#228B22](https://via.placeholder.com/15/228B22/000000?text=+)
+to highlight traffic with "Allow" NetworkPolicy enforced, red color
+![#EE4B2B](https://via.placeholder.com/15/EE4B2B/000000?text=+) to highlight
+traffic with "Deny" NetworkPolicy enforced. Unprotected traffic without NetworkPolicy
+enforced has the same color with its source Pod. Every link is in the shape of
+an arrow, pointing from source to destination. Line graphs shows the evolution
+of traffic throughput with NetworkPolicy enforced.
 
-<img src="https://downloads.antrea.io/static/02152022/flow-visibility-np-1.png" width="900" alt="Network-Policy Flows Dashboard">
+<img src="https://github.com/heanlan/grafana-chord-plugin/releases/download/1.0.0/np-0.png" width="900" alt="Network-Policy Flows Dashboard">
```
Comment (annakhm) [inline docs/network-flow-visibility.md:399]: shouldn't those images be stored under antrea/theia repo (same below)?

Comment (heanlan) [reply docs/network-flow-visibility.md:399]: all of the generated links are temporary. I'll upload those to antrea.io S3 once we finalize the PR.

File: docs/network-flow-visibility.md [hunk 2]
```diff
@@ -387,11 +387,36 @@ and throughput is shown in the line graphs.
...
 
 Network-Policy Flows Dashboard visualizes the traffic with NetworkPolicies enforced.
-Currently we only support the visualization of NetworkPolicies with `Allow` action.
+The Chord diagram visualizes the cumulative bytes of all the traffic in the selected
+time range. We use green color ![#228B22](https://via.placeholder.com/15/228B22/000000?text=+)
+to highlight traffic with "Allow" NetworkPolicy enforced, red color
+![#EE4B2B](https://via.placeholder.com/15/EE4B2B/000000?text=+) to highlight
+traffic with "Deny" NetworkPolicy enforced. Unprotected traffic without NetworkPolicy
+enforced has the same color with its source Pod. Every link is in the shape of
+an arrow, pointing from source to destination. Line graphs shows the evolution
```
Comment (annakhm) [inline docs/network-flow-visibility.md:396]: show

File: build/yamls/base/provisioning/datasources/create_table.sh [hunk 3]
```diff
@@ -163,19 +163,27 @@ clickhouse client -n -h 127.0.0.1 <<-EOSQL
...
 
-    CREATE MATERIALIZED VIEW flows_policy_view
+    CREATE MATERIALIZED VIEW IF NOT EXISTS flows_policy_view
     ENGINE = SummingMergeTree
     ORDER BY (
         timeInserted,
         flowEndSeconds,
         flowEndSecondsFromSourceNode,
         flowEndSecondsFromDestinationNode,
-        egressNetworkPolicyName,
```
Comment (dreamtalen) [inline build/yamls/base/provisioning/datasources/create_table.sh:173]: Why we remove the `egressNetworkPolicyName` column and add `egressNetworkPolicyNamespace` column instead?

File: docs/network-flow-visibility.md [hunk 4]
```diff
@@ -387,11 +387,36 @@ and throughput is shown in the line graphs.
 #### Network-Policy Flows Dashboard
 
 Network-Policy Flows Dashboard visualizes the traffic with NetworkPolicies enforced.
-Currently we only support the visualization of NetworkPolicies with `Allow` action.
+The Chord diagram visualizes the cumulative bytes of all the traffic in the selected
+time range. We use green color ![#228B22](https://via.placeholder.com/15/228B22/000000?text=+)
+to highlight traffic with "Allow" NetworkPolicy enforced, red color
+![#EE4B2B](https://via.placeholder.com/15/EE4B2B/000000?text=+) to highlight
+traffic with "Deny" NetworkPolicy enforced. Unprotected traffic without NetworkPolicy
```
Comment (dreamtalen) [inline docs/network-flow-visibility.md:394]: Will the "Reject" traffic also be highlighted in red?

File: build/yamls/flow-visibility.yml [hunk 5]
```diff
@@ -138,14 +145,17 @@ data:
...
+    \       sourcePodName,\n        sourceTransportPort,\n        sourcePodNamespace,\n
+    \       destinationPodName,\n        destinationTransportPort,\n        destinationPodNamespace,\n
+    \       destinationServicePort,\n        destinationServicePortName,\n        destinationIP;\n\n
     \   CREATE TABLE IF NOT EXISTS recommendations (\n        id String,\n        type
     String,\n        timeCreated DateTime,\n        yamls String\n    ) engine=MergeTree\n
     \   ORDER BY (timeCreated);\n    \nEOSQL\n"
 kind: ConfigMap
 metadata:
-  name: clickhouse-mounted-configmap-dkbmg82ctg
+  name: clickhouse-mounted-configmap-cc8984h57b
```
[the remark this answers is not in the document]
Comment (heanlan) [reply build/yamls/flow-visibility.yml:158]: this is automatically generated by kustomize

**Review by ziyouw — APPROVED** [review]
Thanks for working on this!

File: build/yamls/base/grafana.yml [hunk 6]
```diff
@@ -107,7 +107,7 @@ spec:
           imagePullPolicy: IfNotPresent
           env:
             - name: GF_INSTALL_PLUGINS
-              value: "https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-sankey-plugin-1.0.0.zip;theia-grafana-sankey-plugin,grafana-clickhouse-datasource 1.0.1"
+              value: "https://downloads.antrea.io/artifacts/grafana-custom-plugins/theia-grafana-sankey-plugin-1.0.0.zip;theia-grafana-sankey-plugin,https://github.com/heanlan/grafana-chord-plugin/releases/download/1.0.0/grafana-chord-plugin-1.0.0.zip;theia-grafana-chord-plugin,grafana-clickhouse-datasource 1.0.1"
```
Comment (salv-orlando) [inline build/yamls/base/grafana.yml:110]: Shall we update this before we merge the PR?

**Review by salv-orlando — APPROVED** [review]
LGTM

Let me know how you prefer to go about updating links to the plugin in docs and yaml file.

File: build/yamls/base/grafana.yml (hunk 6 above)
Comment (heanlan) [reply build/yamls/base/grafana.yml]: all the links are updated

**Comment by heanlan** [comment]
@salv-orlando Thanks for the review. All the links in doc and yaml are updated. Would you like to take another look before we merge it?

**Review by salv-orlando — APPROVED** [review]
PR LGTM! Thanks @heanlan

**Pull request merged**